glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt

```diff
@@ -135,82 +137,78 @@ open class GleanInternalAPI internal constructor () {
+            if (isFirstRun) {
+                initializeCoreMetrics(applicationContext)
+            }
+
+            // Deal with any pending events so we can start recording new ones
             val pingSubmitted = LibGleanFFI.INSTANCE.glean_on_ready_to_submit_pings(
                 this@GleanInternalAPI.handle
             ).toBoolean()
             if (pingSubmitted) {
                 PingUploadWorker.enqueueWorker(applicationContext)
             }
-        }
 
-        // Set up information and scheduling for Glean owned pings. Ideally, the "metrics"
-        // ping startup check should be performed before any other ping, since it relies
-        // on being dispatched to the API context before any other metric.
-        metricsPingScheduler = MetricsPingScheduler(applicationContext)
-        metricsPingScheduler.schedule()
+            // Set up information and scheduling for Glean owned pings. Ideally, the "metrics"
+            // ping startup check should be performed before any other ping, since it relies
+            // on being dispatched to the API context before any other metric.
+            metricsPingScheduler = MetricsPingScheduler(applicationContext)
+            metricsPingScheduler.schedule()
 
-        // From the second time we run, after all startup pings are generated,
-        // make sure to clear `lifetime: application` metrics and set them again.
-        // Any new value will be sent in newly generted pings after startup.
-        if (!isFirstRun) {
-            @Suppress("EXPERIMENTAL_API_USAGE")
-            Dispatchers.API.executeTask {
+            // From the second time we run, after all startup pings are generated,
+            // make sure to clear `lifetime: application` metrics and set them again.
+            // Any new value will be sent in newly generted pings after startup.
+            if (!isFirstRun) {
                 LibGleanFFI.INSTANCE.glean_clear_application_lifetime_metrics(handle)
                 initializeCoreMetrics(applicationContext)
             }
-        }
 
-        // Signal Dispatcher that init is complete
-        @Suppress("EXPERIMENTAL_API_USAGE")
-        Dispatchers.API.flushQueuedInitialTasks()
+            // Signal Dispatcher that init is complete
+            Dispatchers.API.flushQueuedInitialTasks()
 
-        // At this point, all metrics and events can be recorded.
-        // This should only be called from the main thread. This is enforced by
-        // the @MainThread decorator and the `assertOnUiThread` call.
-        ProcessLifecycleOwner.get().lifecycle.addObserver(gleanLifecycleObserver)
+            // At this point, all metrics and events can be recorded.
+            // This should only be called from the main thread. This is enforced by
+            // the @MainThread decorator and the `assertOnUiThread` call.
+            MainScope().launch {
+                ProcessLifecycleOwner.get().lifecycle.addObserver(gleanLifecycleObserver)
+            }
```

Comment:
Do we need to do this here, or could we do it outside of the coroutine (near the top where things like `applicationContext` are set?)

Reply:
I'm not sure yet, I don't think there's anything preventing us from moving it there, before the glean init even starts.

The only two reasons for not doing that would be:

- we don't want to register this if the glean init fails;
- we really want to test this in Fenix before making a call!

Reply:
Fair enough on both points.  I think I was just coming at it from the angle of -- going off the main thread only to come back to it is more complicated.  But, as you say, we don't really know until we try it in the context of Fenix.

Reply:
I'll make sure to try that first once I get rid of the failing test and have a proper Fenix build for testing!

Reply:
I tested that, but I think we should retain the current behaviour even just to make sure that we don't mistakenly register observers if init fails. Thoughts?

Reply:
That's a good point about not registering if we exit early for some other error during init.  :+1: on this, then!